server/datastore/mysql/software.go

```diff
@@ -519,3 +533,82 @@ func (d *Datastore) SoftwareByID(ctx context.Context, id uint) (*fleet.Software,
 
 	return &software, nil
 }
+
+// CalculateHostsPerSoftware calculates the number of hosts having each
+// software installed and stores that information in the aggregated_stats
+// table.
+func (d *Datastore) CalculateHostsPerSoftware(ctx context.Context, updatedAt time.Time) error {
+	// NOTE(mna): for reference, on my laptop I get ~1.5ms for 10_000 hosts / 100 software each,
+	// ~1.5s for 10_000 hosts / 1_000 software each (but this is with an otherwise empty
+	// aggregated_stats table, but still reasonable numbers give that this runs as a cron
+	// task in the background).
+
+	resetStmt := `
+    UPDATE aggregated_stats
+    SET json_value = CAST(0 AS json)
+    WHERE type = "software_hosts_count"`
+
+	queryStmt := `
+    SELECT count(*), software_id
+    FROM host_software
+    GROUP BY software_id`
+
+	insertStmt := `
+    INSERT INTO aggregated_stats
+      (id, type, json_value, updated_at)
+    VALUES
+      %s
+    ON DUPLICATE KEY UPDATE
+      json_value = VALUES(json_value),
+      updated_at = VALUES(updated_at)`
+	valuesPart := `(?, "software_hosts_count", CAST(? AS json), ?),`
+
+	// first, reset all counts to 0
+	if _, err := d.writer.ExecContext(ctx, resetStmt); err != nil {
```

Comment:
Something to keep in mind here - if the DB went down after this, then the counts would all be 0, until the next time this runs (the vuln interval). I think this is fine as the UI also supports this elegantly (based on the figma wireframes); another approach would be to use a transaction around the reset/inserts, but I'm afraid this potentially is a lot of data to put together in a transaction.

Another thing I wanted to mention - we could take advantage of this new "background" calculation of counts to clean-up unused software entries (which is a TODO elsewhere in this file) - after the inserts, any software with a count of 0 is unused (or to take it the other way around, which would catch software that never had their counts calculated, any software that doesn't have an entry with a count > 0 is unused).

Reply:
> Something to keep in mind here - if the DB went down after this, then the counts would all be 0, until the next time this runs (the vuln interval). I think this is fine as the UI also supports this elegantly (based on the figma wireframes); another approach would be to use a transaction around the reset/inserts, but I'm afraid this potentially is a lot of data to put together in a transaction.

Agreed this is not ideal, but doing it in a tx might have other performance implications that I'm not sure we want to deal with at this point.

> Another thing I wanted to mention - we could take advantage of this new "background" calculation of counts to clean-up unused software entries (which is a TODO elsewhere in this file) - after the inserts, any software with a count of 0 is unused (or to take it the other way around, which would catch software that never had their counts calculated, any software that doesn't have an entry with a count > 0 is unused).

Yeah, cleaning up makes sense to me.

Reply:
I'll wait for Lucas' review and clarification on the software inventory / vulnerabilities config comment from Lucas, but otherwise I'll merge as-is and create another PR to add the unused software cleanup.